File: ruby/reverse_string.rb
def reverse_string(str)
  # type your code in here
  reversed_string = []
  str.length.times do |i|
    reversed_string << str[str.length-1]
    str = str.chop
  end
  reversed_string.join
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'ih'"
  puts "=>", reverse_string('hi')

  puts

  puts "Expecting: 'ybabtac'"
  puts "=>", reverse_string('catbaby')

  puts "Expecting: 'sailE'"
  puts "=>", reverse_string('Elias')

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

#Pseudocode

# define method that accepts a string as an argument
# define new_string variable set to an empty array

# iterate over each character in the string
# add last character from string to  the empty array
#delete last character from the string
# join the letters in array to a string

# return new_string in reverse order

#Explanation : First you have to define a reversed_string Array. Than when you iterate
#over the string grab the last character of the original string and add it to the reversed_string array.
#Then you have to use .chop to remove the last character from the string and mutate the string.
#Now you have an array of letters in reverse order so you use .join to join the letters and return it as
#a reversed string.

File: javascript/reverse_string.js
function reverseString(str) {
  let reversedString = []
  let splitString = str.split("")
  let i = 1

  for (const letter of splitString) {
    let letterToAdd = splitString[splitString.length-i]
    reversedString.push(letterToAdd)
    i++
  }

  return reversedString.join("")
}



if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'ih'");
  console.log("=>", reverseString("hi"));

  console.log("");

  console.log("Expecting: 'ybabtac'");
  console.log("=>", reverseString("catbaby"));

  console.log("Expectin: 'sailE");
  console.log("=>", reverseString("Elias"));
}

module.exports = reverseString;


// Please add your pseudocode to this file
// And a written explanation of your solution


// PseudoCode

// define function that accepts a string as an argument
// define reversed_String variable set to an empty array
// definie index variable set to 1
// split string into array of letters

// loop through each character in the string
// iterte through each letter in reverse by increasing the index and assign letter to a variable
// add variable from string to  the empty array
// join the letters in array to a string

// return reversed_string in reverse order

// If you iterate over the string, then take each character and bring it to the front
// of the string then the word will be reversed. We can do this by collecting all the
// letters in reverse order and then join those letters into a string and then returning
//  that string.